### PR TITLE
nixos-rebuild: run activation inside systemd-run

### PR DESCRIPTION
**Pull request by thiagokokada**
## Description of changes

Right now, running `nixos-rebuild switch` in a remote system via SSH may put the system in an unusable state by restarting services (e.g.: network ones like systemd-networkd.service) during the update. This will cause the SSH to lose access to the TTY, stopping the process entirely.

This commit wraps up the call to `switch-to-configuration` inside a systemd-run call, making the switch resiliant against TTY loss, since it will allocate its own TTY. For the user this should be entirely transparent though, with the only visible change is that the `switch-to-configuration` messages will be logged through journalctl (e.g.: `journalctl -u nixos-rebuild-switch-to-configuration`).

Fix issue https://github.com/NixOS/nixpkgs/issues/39118.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
For testers: you will need to set my branch as your `nixpkgs` source, because `nixos-rebuild` will exec the `nixos-rebuild` from your nixpkgs configuration instead of using the original script.

To check if everything is working, running a `sudo nixos-rebuild switch` and checking `journalctl -u nixos-rebuild-switch-to-configuration` should be sufficient.

**Comment by thiagokokada** [comment]
CC @K900 @NickCao @roberth @Profpatsch @jonringer.

**Comment by K900** [comment]
First question: will this work in a chroot?

**Comment by thiagokokada** [comment]
> First question: will this work in a chroot?

Not sure how to test this. Is there any tests that we can run?

**Comment by K900** [comment]
`nixos-enter` a system and try to `nixos-rebuild boot`

**Comment by thiagokokada** [comment]
Tested and working fine too:

```
nixos-rebuild switch --flake /etc/nixos#<hostname> --target-host root@<hostname> --use-substitutes
```

**Comment by thiagokokada** [comment]
@ofborg build nixos-rebuild

@ofborg test installer misc nixos-rebuild-specialisations

**Comment by thiagokokada** [comment]
Fixed the logic for the specialisations case:

@ofborg build nixos-rebuild

@ofborg test installer misc nixos-rebuild-specialisations

**Comment by ElvishJerricco** [comment]
@thiagokokada `nixos-enter --root /` is not a useful test. I think @K900 wanted to know something more like what happens when you mount the file system of another system that has this patch and try to `nixos-enter` and `nixos-rebuild` it. I expect it won't work because there's no systemd running in that chroot to handle the `systemd-run`.

**Comment by Ma27** [comment]
fyi this was tried for nspawn containers in the past and had to be reverted because it broke on reloading systemd itself: https://github.com/NixOS/nixpkgs/pull/83392

Not sure if that's still an issue, but you may want to check for this.

**Comment by thiagokokada** [comment]
> @thiagokokada `nixos-enter --root /` is not a useful test. I think @K900 wanted to know something more like what happens when you mount the file system of another system that has this patch and try to `nixos-enter` and `nixos-rebuild` it. I expect it won't work because there's no systemd running in that chroot to handle the `systemd-run`.

In this case yes, won't work but this is pretty obvious. I was more of the impression that the issue here is that @K900 wanted to test is if `systemd-run` would work inside a chroot environment with NixOS (e.g.: NixOS live), not in any kind of chroot environment (e.g.: a different distro that does not support systemd).

In this particular case that you described, I think I can check if `systemd-run` is available in the environment and a fallback to not use `systemd-run` in this case. WDYT?

**Comment by ElvishJerricco** [comment]
> I was more of the impression that the issue here is that @K900 wanted to test is if systemd-run would work inside a chroot environment with NixOS (e.g.: NixOS live), not in any kind of chroot environment (e.g.: a different distro that does not support systemd).

No, I'm not concerned with other distros; running nixos-rebuild on another distro is meaningless. When you chroot to a NixOS file system and run `systemd-run`, it will not work, because there is no systemd running in that chroot. The `systemd-run` command will fail to connect to a system manager and it will fail.

**Comment by thiagokokada** [comment]
> Not sure if that's still an issue, but you may want to check for this.

Not sure exactly how to test for this issue. It seems that running `systemd-run -t --wait bash` and trying `systemctl daemon-reload` or `systemctl daemon-reexec` doesn't cause the `systemd-run` shell to die, but maybe there is a better way to test this.

**Comment by thiagokokada** [comment]
> > I was more of the impression that the issue here is that @K900 wanted to test is if systemd-run would work inside a chroot environment with NixOS (e.g.: NixOS live), not in any kind of chroot environment (e.g.: a different distro that does not support systemd).
> 
> No, I'm not concerned with other distros; running nixos-rebuild on another distro is meaningless. When you chroot to a NixOS file system and run `systemd-run`, it will not work, because there is no systemd running in that chroot. The `systemd-run` command will fail to connect to a system manager and it will fail.

Just tested and it really is not working inside the Live NixOS environment. I will add a check testing `systemd-run` first and if it fails it will fallback to run it without `systemd-run`.

**Comment by thiagokokada** [comment]
> No, I'm not concerned with other distros; running nixos-rebuild on another distro is meaningless. When you chroot to a NixOS file system and run `systemd-run`, it will not work, because there is no systemd running in that chroot. The `systemd-run` command will fail to connect to a system manager and it will fail.

Tested after https://github.com/NixOS/nixpkgs/pull/258571/commits/2f6a7da9b111a9c617f0556f805ac694612a78a8:
- Newly created VM in Gnome Boxes with NixOS Live ISO
- Created a Ext4 filesystem, mounted it on `/mnt`
- `nixos-generate-config --root /mnt`
- `nixos-install --root /mnt`
- `nixos-enter` and `nixos-rebuild switch --verbose`

Result: the switch works and I can see the newly "Skipping systemd-run since it is not working in target host" message. It will fail later because of `busctl`, but this is unrelated to this branch because I can reproduce the same error in `master`.

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:662]: `switch-to-configuration` should not use `NIX_PATH` because everything has already been built. At this point nothing should need to call Nix commands anymore.

```suggestion
```

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:663]: This value comes from the old system, not the new system. I wonder if that's intentional...
(but not a regression fwiw)

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:661]:
```suggestion
        "-E" "LOCALE_ARCHIVE"  # Will be set to new value early in switch-to-configuration script, but interpreter starts out with old value
```

**Review by roberth — COMMENTED** [review]
Thoughts about old system environment "leaking" into the new system. This PR already creates a cleaner environment, so that's probably good.

**Comment by roberth** [comment]
This runs the switch script in a cleaner environment, which *may* break scripts that relied on the dirty environment that we used to have. Could you add a release note about this?

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:662]: Good point. I actually extract those values from my `nixos-rebuild` wrapper (from #39118), and it needs there because of the build phase. But in this case not.

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:663]: I think this is also not actually needed, let me double check.

Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:663]: Yes, not needed. Will remove it too.

**Comment by thiagokokada** [comment]
> This runs the switch script in a cleaner environment, which _may_ break scripts that relied on the dirty environment that we used to have. Could you add a release note about this?

Done.

**Comment by thiagokokada** [comment]
> This runs the switch script in a cleaner environment, which _may_ break scripts that relied on the dirty environment that we used to have. Could you add a release note about this?

I added a new environment variable `NIXOS_IGNORE_SYSTEMD_RUN` that once set will skip `systemd-run` (the same as when we find that `systemd-run` is not working). I think this should cover the cases where someone may want to keep their environment "dirty" during the switch.

WDYT @roberth ?

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:680]: I think this should be a warning, because this option is not sustainable in the long run.
We can burden neither users nor maintainers with this distinction, so it should only exist as an escape hatch for a limited time. (A year?)
Scripts that relied on this should be fixed.

Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: Done.

**Comment by thiagokokada** [comment]
> Commit message typo:
> 
> > release-nodes: document nixos-rebuild changes
> 
> Should be `release-notes` :-)

Thank you, fixed.

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (ElvishJerricco) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: @roberth What do you mean? We *can't* use `systemd-run` in a chroot, so this has to stay if we want to be able to `nixos-rebuild` in a `nixos-enter`.

Comment (roberth) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: @ElvishJerricco you're right!

The warning should only apply when `NIXOS_IGNORE_SYSTEMD_RUN` is set.

Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: Fixed @roberth @ElvishJerricco.

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:680]: When using `--target`, this is an extra ssh roundtrip that could be avoided if we make `switch-to-configuration` responsible for re-execing itself in `systemd-run` - ie move all the new logic into that command (or a custom wrapper for it).

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:682]: In this case we should clean the environment, for consistency with normal deployments.

`NIXOS_IGNORE_SYSTEMD_RUN` should perhaps be `NIXOS_SWITCH_USE_DIRTY_ENV`.

Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:682]: Done.

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:680]: I took a look at the `switch-configuration` and it seems that this is in Perl and I am not really comfortable in modifying it.

Is this extra round-trip a blocker for this PR? Or could we leave this as-is and change it later?

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (roberth) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:682]: I think we should clean the environment for consistency with `systemd-run`.
`switch-to-configuration` will set `PATH`.

A first approximation would be

```suggestion
        cmd=("env" "-i" "LOCALE_ARCHIVE=$LOCALE_ARCHIVE")
```

However, I don't know what `systemd-run` preserves and/or loads.

Comment (thiagokokada) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: Done.

**Review by thiagokokada — COMMENTED** [review]
CC @roberth @K900 @ElvishJerricco, can I ask a re-review?

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh
Comment (Ma27) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: fwiw unless you've done `systemctl import-environment` before. For instance

```
$ export FNORD=bar
$ sudo -E systemctl import-environment FNORD
$ sudo systemd-run -t --wait env 2>/dev/null | grep FNORD
FNORD=bar
```

Also, `import-environment` w/o args imports the full environment.

**Comment by roberth** [comment]
I've kicked off evaluations on hydra to be sure.
- [baseline](https://hydra.nixos.org/eval/1800086) right before this PR at 3d007e8a60c4d1d4ea02ab9912573fb839ef6c88
- _(number and link tbd)_ this PR at 1e79cca679380dd9fa05e76c11a22815cfcde773
- _(tbd comparison link)_

**Comment by roberth** [comment]
> [comparison](https://hydra.nixos.org/eval/1800094?compare=1800086)

None of the _newly failed_ seem to use configuration switching, so those all seem to be flaky tests.

**Review by roberth — APPROVED** [review]

**Comment by thiagokokada** [comment]
Thanks @roberth. Would like one more reviewer here.

CC @K900 @ElvishJerricco @NickCao @Profpatsch @jonringer @bjornfor.

**Comment by thiagokokada** [comment]
If there no other concerns until the end of this week I will merge this PR.

**Comment by thiagokokada** [comment]
Merging it.

**Pull request merged**
